File: global-apps/global-apps-common/src/main/java/io/global/ot/name/NameOTSystem.java
package io.global.ot.name;

import io.datakernel.ot.OTSystem;
import io.datakernel.ot.OTSystemImpl;

import static io.datakernel.ot.TransformResult.*;
import static io.datakernel.util.Preconditions.checkState;
import static io.global.ot.name.ChangeName.changeName;
import static java.util.Collections.singletonList;

public final class NameOTSystem {
	private NameOTSystem() {
		throw new AssertionError();
	}

	public static OTSystem<ChangeName> createOTSystem() {
		return OTSystemImpl.<ChangeName>create()
				.withEmptyPredicate(ChangeName.class, ChangeName::isEmpty)
				.withInvertFunction(ChangeName.class, op -> singletonList(changeName(op.getNext(), op.getPrev(), op.getTimestamp())))

				.withSquashFunction(ChangeName.class, ChangeName.class, (first, second) -> changeName(first.getPrev(), second.getNext(), second.getTimestamp()))

				.withTransformFunction(ChangeName.class, ChangeName.class, (left, right) -> {
					checkState(left.getPrev().equals(right.getPrev()), "Previous values of left and right operations should be equal");
					if (left.getTimestamp() > right.getTimestamp()) {
						return right(changeName(right.getNext(), left.getNext(), left.getTimestamp()));
					}
					if (left.getTimestamp() < right.getTimestamp()) {
						return left(changeName(left.getNext(), right.getNext(), right.getTimestamp()));
					}
					if (left.getNext().compareTo(right.getNext()) > 0) {
						return right(changeName(right.getNext(), left.getNext(), left.getTimestamp()));
					}
					if (left.getNext().compareTo(right.getNext()) < 0) {
						return left(changeName(left.getNext(), right.getNext(), right.getTimestamp()));
					}
					return empty();
				});
	}
}

File: global-apps/global-documents/front/src/modules/documents/ot/name/nameOTSystem.js
import {OTSystemBuilder, TransformResult} from 'ot-core';
import ChangeNameOperation from "./ChangeNameOperation";

const nameOTSystem = new OTSystemBuilder()
  .withEmptyPredicate(ChangeNameOperation, operation => operation.isEmpty())
  .withInvertFunction(ChangeNameOperation, operation => operation.invert())
  .withTransformFunction(ChangeNameOperation, ChangeNameOperation, (left, right) => {
    if (left.timestamp > right.timestamp)
      return TransformResult.right([new ChangeNameOperation(right.next, left.next, left.timestamp)]);
    if (left.timestamp < right.timestamp)
      return TransformResult.left([new ChangeNameOperation(left.next, right.next, right.timestamp)]);
    if (left.next > right.next)
      return TransformResult.right(new ChangeNameOperation(right.next, left.next, left.timestamp));
    if (left.next < right.next)
      return TransformResult.left(new ChangeNameOperation(left.next, right.next, right.timestamp));
    return TransformResult.empty();
  })
  .withSquashFunction(ChangeNameOperation, ChangeNameOperation, (first, second) =>
    new ChangeNameOperation(first.prev, second.next, second.timestamp))
  .build();

export default nameOTSystem;
